### ISSUE TITLE
azurerm_storage_share_file empties files on change

### ISSUE THREAD

**Issue by mgabathuler** [hashicorp/terraform-provider-azurerm#11507]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
```
bash-5.1# terraform -v
Terraform v0.14.9
+ provider registry.terraform.io/hashicorp/azurerm v2.56.0
```

### Affected Resource(s)
* `azurerm_storage_share_file`

### Terraform Configuration Files
```hcl
data "azurerm_storage_account" "plans" {
  name                     = local.env_config.storage.account_name
  resource_group_name      = local.env_config.storage.account_rg
}

resource "azurerm_storage_share_file" "plans" {
  for_each = fileset(path.root, "../../plans/*.json")

  name             = basename(each.key)
  storage_share_id = join("", [
    data.azurerm_storage_account.plans.primary_file_endpoint,
    "plans"
  ])

  source           = each.key
  content_md5 = filemd5(each.key)
}
```

### Expected Behaviour
i would expect the file on the file share to be recreated and file contents to be updated.

documentation says that a change in the content_md5 would cause the resource to be recreated but that does not seem to happen...

### Actual Behaviour
the files do not get recreated. instead files are suddenly empty!

from the terraform plan...
```
Terraform will perform the following actions:

  # azurerm_storage_share_file.plans["../../plans/dqr_plan.json"] will be updated in-place
  ~ resource "azurerm_storage_share_file" "plans" {
      + content_md5      = "ba3d9dc04cf9e15e511781745c52b7a3"
        id               = "https://iv2***jobrunner.file.core.windows.net/plans//dqr_plan.json"
        name             = "dqr_plan.json"
        # (4 unchanged attributes hidden)
    }

  # azurerm_storage_share_file.plans["../../plans/igh_plan.json"] will be updated in-place
  ~ resource "azurerm_storage_share_file" "plans" {
      + content_md5      = "b7a97b5d3b723b036707f8095b1d24f5"
        id               = "https://iv2***jobrunner.file.core.windows.net/plans//igh_plan.json"
        name             = "igh_plan.json"
        # (4 unchanged attributes hidden)
    }

Plan: 0 to add, 3 to change, 0 to destroy.
```

from the apply...
```
azurerm_storage_share_file.plans["../../plans/igh_plan.json"]: Modifying... [id=https://iv2***jobrunner.file.core.windows.net/plans//igh_plan.json]
azurerm_storage_share_file.plans["../../plans/dqr_plan.json"]: Modifying... [id=https://iv2***jobrunner.file.core.windows.net/plans//dqr_plan.json]
azurerm_resource_group.service: Modifying... [id=/subscriptions/***/resourceGroups/iv2-***-***-job-plans]
azurerm_resource_group.service: Modifications complete after 0s [id=/subscriptions/***/resourceGroups/iv2-***-***-job-plans]
azurerm_storage_share_file.plans["../../plans/igh_plan.json"]: Modifications complete after 0s [id=https://iv2***jobrunner.file.core.windows.net/plans//igh_plan.json]
azurerm_storage_share_file.plans["../../plans/dqr_plan.json"]: Modifications complete after 0s [id=https://iv2***jobrunner.file.core.windows.net/plans//dqr_plan.json]

Apply complete! Resources: 0 added, 3 changed, 0 destroyed.
```

looking at the files in the storage account in azure portal
![image](https://user-images.githubusercontent.com/6018698/116385773-6eb5f780-a819-11eb-859a-edbee48791bd.png)

**Comment by siefgred** [comment]
I have the same issue. Any solution found ?

**Comment by TobiTh** [comment]
The content of the file also gets cleared when the content_type property is changed.

**Comment by richard1927** [comment]
I also encountered this issue using both azurerm version 2.42.0 and 2.66.0.

I was using the following Terraform configuration:
```
resource "azurerm_storage_share_file" "rush-hour-load-test" {
  name             = "rush-hour-load-test.js"
  storage_share_id = azurerm_storage_share.k6-shared.id
  source           = "k6/rush-hour-load-test.js"
  content_md5      = filemd5("k6/rush-hour-load-test.js")
}
```

**Comment by drummer3333** [comment]
Same here with azurerm version 2.65.0 or 2.79.1

```
resource "azurerm_storage_share_file" "systems_config" {
  name             = "systems.json"
  storage_share_id = azurerm_storage_share.store_share.id
  source           = "../../systems.json"
  content_md5      = filemd5("../../systems.json")
}
```
as the content_md5 changes, terraform will update md5 value in-place instead of recreating the resource as stated in the documentation. After that the file content is 0 bytes.

As a workaround im using `-replace="azurerm_storage_share_file.systems_config"` with every `terraform apply`.

**Comment by smnmtzgr** [comment]
@tombuildsstuff  any news here?

**Comment by berndos** [comment]
:+1:

**Comment by fwiedmann** [comment]
I looked into the code and as far I understand that there are three different kind of bugs:

- The the documentation states that after the content of `content_md5` changes the file should be recreated. But the  `schema` config does not have the flag `ForceNew` (https://github.com/hashicorp/terraform-provider-azurerm/blob/v2.92.0/internal/services/storage/storage_share_file_resource.go#L68)
- The ID of the resource has always a duplicated `/`.   As you can see in the bug report
```bash
id=https://iv2***jobrunner.file.core.windows.net/plans//dqr_plan.json
```
- The `resourceStorageShareFileUpdate` function does check for changes on the resource, but should not check on `content_md5` due to the file should be recreated (as the docs states). Furthermore the `client.SetProperties` will delete the files content. 

@tombuildsstuff  I can look into this to fix :)

**Comment by fwiedmann** [comment]
@smnmtzgr Fix is merged, will be in the next release I guess 🎊

**Comment by rcskosir** [comment]
Thanks for taking the time to submit this issue. It looks like this has been resolved as of #15007. As such, I am going to mark this issue as closed.
